waiter/src/waiter/scheduler.clj

```diff
@@ -391,54 +392,121 @@
+    (log/trace "scheduler-syncer: querying scheduler")
+    (if-let [service->service-instances (timers/start-stop-time!
+                                          (metrics/waiter-timer "core" "scheduler" "app->available-tasks")
+                                          (do-health-checks (request-available-waiter-apps scheduler)
+                                                            (fn available [instance health-check-path]
+                                                              (available? instance health-check-path http-client))
+                                                            service-id->service-description-fn))]
+      (let [available-services (keys service->service-instances)
+            available-service-ids (map :id available-services)]
+        (log/debug "scheduler-syncer:" (count service->service-instances) "available services:" available-service-ids)
+        (doseq [service available-services]
+          (when (zero? (reduce + 0 (filter number? (vals (select-keys (:task-stats service) [:staged :running :healthy :unhealthy])))))
+            (log/info "scheduler-syncer:" (:id service) "has no live instances!" (:task-stats service))))
+        (loop [service-id->health-check-context' {}
+               scheduler-messages [[:update-available-apps {:available-apps available-service-ids :scheduler-sync-time request-apps-time}]]
+               [[service {:keys [active-instances failed-instances]}] & remaining] (seq service->service-instances)]
+          (if service
+            (let [request-instances-time (t/now)
+                  service-id (:id service)
+                  {:keys [instance-id->unhealthy-instance
+                          instance-id->tracked-failed-instance
+                          instance-id->failed-health-check-count]} (get service-id->health-check-context service-id)
+                  service-instance-info (retrieve-instances-for-app service-id active-instances)
+                  instance-id->unhealthy-instance' (into {} (map #(vector (:id %) %)
+                                                                 (:unhealthy-instances service-instance-info)))
+                  not-in? (fn [id instances] (not (some #(= id (:id %)) instances)))
+                  instance-id->tracked-failed-instance' ((fnil into {}) instance-id->tracked-failed-instance
+                                                          (map #(vector % (get instance-id->unhealthy-instance %))
+                                                               (filter #(and (not-in? % active-instances)
+                                                                             (> (get instance-id->failed-health-check-count %) max-failed-health-checks))
+                                                                       (keys instance-id->unhealthy-instance))))
+                  failed-instances (into failed-instances
```

Comment:
It's better to add the `:never-passed-health-check` flag when you're adding the instance to the `instance-id->tracked-failed-instance`' map so you don't have to add it here.